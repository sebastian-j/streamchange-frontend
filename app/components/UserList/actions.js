/*
 *
 * UserList actions
 *
 */

import {
  DESELECT_ALL,
  GET_LIST_FROM_IDB,
  PURGE_LIST,
  PUSH_USER,
  SELECT_ALL,
  TOGGLE_ELIGIBILITY,
  UPDATE_MESSAGE,
} from './constants';

export function selectAllUsers() {
  return {
    type: SELECT_ALL,
  };
}

export function getListFromIdb(arr) {
  return {
    type: GET_LIST_FROM_IDB,
    userArray: arr,
  };
}

export function purgeList() {
  return {
    type: PURGE_LIST,
  };
}

export function pushUser(item) {
  return {
    type: PUSH_USER,
    item,
  };
}

export function deselectAllUsers() {
  return {
    type: DESELECT_ALL,
  };
}

export function toggleEligibility(userId) {
  return {
    type: TOGGLE_ELIGIBILITY,
    userId,
  };
}

export function updateMessage(item) {
  return {
    type: UPDATE_MESSAGE,
    item,
  };
}
