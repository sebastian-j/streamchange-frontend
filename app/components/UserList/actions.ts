/*
 *
 * UserList actions
 *
 */

import { action } from 'typesafe-actions';
import ActionTypes from './constants';
import { User } from './types';

export const selectAllUsers = () => action(ActionTypes.SELECT_ALL);

export const getListFromIdb = (userArray: User[]) =>
  action(ActionTypes.GET_LIST_FROM_IDB, userArray);

export const purgeList = () => action(ActionTypes.PURGE_LIST);

export const pushUser = (item: User) => action(ActionTypes.PUSH_USER, item);

export const deselectAllUsers = () => action(ActionTypes.DESELECT_ALL);

export const toggleEligibility = (userId: string) =>
  action(ActionTypes.TOGGLE_ELIGIBILITY, userId);
