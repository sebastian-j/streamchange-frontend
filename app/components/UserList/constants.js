/*
 *
 * UserList constants
 *
 */

export const DESELECT_ALL = 'app/UserList/DESELECT_ALL';
export const GET_LIST_FROM_IDB = 'app/UserList/GET_LIST_FROM_IDB';
export const PURGE_LIST = 'app/UserList/PURGE_LIST';
export const PUSH_USER = 'app/UserList/PUSH_USER';
export const SELECT_ALL = 'app/UserList/SELECT_ALL';
export const TOGGLE_ELIGIBILITY = 'app/UserList/TOGGLE_ELIGIBILITY';
export const UPDATE_MESSAGE = 'app/UserList/UPDATE_MESSAGE';
