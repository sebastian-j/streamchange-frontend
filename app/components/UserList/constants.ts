/*
 *
 * UserList constants
 *
 */

enum ActionTypes {
  DESELECT_ALL = 'app/UserList/DESELECT_ALL',
  GET_LIST_FROM_IDB = 'app/UserList/GET_LIST_FROM_IDB',
  PURGE_LIST = 'app/UserList/PURGE_LIST',
  PUSH_USER = 'app/UserList/PUSH_USER',
  SELECT_ALL = 'app/UserList/SELECT_ALL',
  TOGGLE_ELIGIBILITY = 'app/UserList/TOGGLE_ELIGIBILITY',
}

export default ActionTypes;
