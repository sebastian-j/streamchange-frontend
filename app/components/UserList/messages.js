/*
 * UserList Messages
 *
 * This contains all the text for the UserList component.
 */
import { defineMessages } from 'react-intl';

export const scope = 'streamchange.components.UserList';

export default defineMessages({
  panelTitle: {
    id: `${scope}.title`,
    defaultMessage: 'Users',
  },
  searchPlaceholder: {
    id: `${scope}.search.placeholder`,
    defaultMessage: 'Search',
  },
  clearBtn: {
    id: `${scope}.clearBtn`,
    defaultMessage: 'Clear all',
  },
  counter: {
    id: `${scope}.counter`,
    defaultMessage: '{selected} out of {all} are participating',
  },
});
