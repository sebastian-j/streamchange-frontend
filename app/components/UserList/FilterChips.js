import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { FormattedMessage, useIntl } from 'react-intl';
import Avatar from '@mui/material/Avatar';
import Chip from '@mui/material/Chip';

import messages from './messages';

const Title = styled.span`
  color: ${(props) => props.theme.staticTextColor};
`;

const ListItem = styled.li`
  display: inline-block;
  list-style: none;
  margin: 0 0 5px 7px;
  .MuiChip-root {
    background-color: #e0e0e0;
    .MuiAvatar-root {
      background-color: #f1f1f1;
    }
  }
`;

const FilterChips = (props) => {
  const intl = useIntl();
  const isFiltering = () => !Object.values(props.filters).some((x) => x);
  return (
    <div>
      {!isFiltering() && (
        <div>
          <Title>
            <FormattedMessage {...messages.filtersTooltip} />
          </Title>
          {props.filters.regulars && (
            <ListItem>
              <Chip
                avatar={
                  <Avatar>
                    <svg
                      xmlns="http://www.w3.org/2000/svg"
                      height="24px"
                      viewBox="0 0 24 24"
                      width="24px"
                      fill="rgb(130,55,238)"
                    >
                      <path d="M0 0h24v24H0V0z" fill="none" />
                      <path d="M10.25 13c0 .69-.56 1.25-1.25 1.25S7.75 13.69 7.75 13s.56-1.25 1.25-1.25 1.25.56 1.25 1.25zM15 11.75c-.69 0-1.25.56-1.25 1.25s.56 1.25 1.25 1.25 1.25-.56 1.25-1.25-.56-1.25-1.25-1.25zm7 .25c0 5.52-4.48 10-10 10S2 17.52 2 12 6.48 2 12 2s10 4.48 10 10zM10.66 4.12C12.06 6.44 14.6 8 17.5 8c.46 0 .91-.05 1.34-.12C17.44 5.56 14.9 4 12 4c-.46 0-.91.05-1.34.12zM4.42 9.47c1.71-.97 3.03-2.55 3.66-4.44C6.37 6 5.05 7.58 4.42 9.47zM20 12c0-.78-.12-1.53-.33-2.24-.7.15-1.42.24-2.17.24-3.13 0-5.92-1.44-7.76-3.69C8.69 8.87 6.6 10.88 4 11.86c.01.04 0 .09 0 .14 0 4.41 3.59 8 8 8s8-3.59 8-8z" />
                    </svg>
                  </Avatar>
                }
                label={intl.formatMessage({...messages.regularsFilter})}
                onDelete={() => props.onDelete('regulars')}
              />
            </ListItem>
          )}
          {props.filters.moderators && (
            <ListItem>
              <Chip
                avatar={
                  <Avatar>
                    <svg viewBox="0 0 16 16" fill="rgb(94, 132, 241)">
                      <g>
                        <path d="M9.64589146,7.05569719 C9.83346524,6.562372 9.93617022,6.02722257 9.93617022,5.46808511 C9.93617022,3.00042984 7.93574038,1 5.46808511,1 C4.90894765,1 4.37379823,1.10270499 3.88047304,1.29027875 L6.95744681,4.36725249 L4.36725255,6.95744681 L1.29027875,3.88047305 C1.10270498,4.37379824 1,4.90894766 1,5.46808511 C1,7.93574038 3.00042984,9.93617022 5.46808511,9.93617022 C6.02722256,9.93617022 6.56237198,9.83346524 7.05569716,9.64589147 L12.4098057,15 L15,12.4098057 L9.64589146,7.05569719 Z" />
                      </g>
                    </svg>
                  </Avatar>
                }
                label={intl.formatMessage({...messages.moderatorsFilter})}
                onDelete={() => props.onDelete('moderators')}
              />
            </ListItem>
          )}
          {props.filters.sponsors && (
            <ListItem>
              <Chip
                avatar={
                  <Avatar>
                    <svg
                      xmlns="http://www.w3.org/2000/svg"
                      enableBackground="new 0 0 24 24"
                      height="24px"
                      viewBox="0 0 24 24"
                      width="24px"
                      fill="rgb(43,166,64)"
                    >
                      <g>
                        <rect fill="none" height="24" width="24" />
                      </g>
                      <g>
                        <path d="M12,2C6.48,2,2,6.48,2,12s4.48,10,10,10s10-4.48,10-10S17.52,2,12,2z M12,20c-4.41,0-8-3.59-8-8c0-4.41,3.59-8,8-8 s8,3.59,8,8C20,16.41,16.41,20,12,20z M12.89,11.1c-1.78-0.59-2.64-0.96-2.64-1.9c0-1.02,1.11-1.39,1.81-1.39 c1.31,0,1.79,0.99,1.9,1.34l1.58-0.67c-0.15-0.44-0.82-1.91-2.66-2.23V5h-1.75v1.26c-2.6,0.56-2.62,2.85-2.62,2.96 c0,2.27,2.25,2.91,3.35,3.31c1.58,0.56,2.28,1.07,2.28,2.03c0,1.13-1.05,1.61-1.98,1.61c-1.82,0-2.34-1.87-2.4-2.09L8.1,14.75 c0.63,2.19,2.28,2.78,3.02,2.96V19h1.75v-1.24c0.52-0.09,3.02-0.59,3.02-3.22C15.9,13.15,15.29,11.93,12.89,11.1z" />
                      </g>
                    </svg>
                  </Avatar>
                }
                label={intl.formatMessage({...messages.sponsorsFilter})}
                onDelete={() => props.onDelete('sponsors')}
              />
            </ListItem>
          )}
          {props.filters.verified && (
            <ListItem>
              <Chip
                avatar={
                  <Avatar>
                    <svg
                      xmlns="http://www.w3.org/2000/svg"
                      height="24px"
                      viewBox="0 0 24 24"
                      width="24px"
                      fill="#ffad29"
                    >
                      <path d="M0 0h24v24H0V0z" fill="none" />
                      <path d="M9 16.2L4.8 12l-1.4 1.4L9 19 21 7l-1.4-1.4L9 16.2z" />
                    </svg>
                  </Avatar>
                }
                label={intl.formatMessage({...messages.verifiedFilter})}
                onDelete={() => props.onDelete('verified')}
              />
            </ListItem>
          )}
          {props.filters.participating && (
            <ListItem>
              <Chip
                avatar={
                  <Avatar>
                    <svg
                      xmlns="http://www.w3.org/2000/svg"
                      height="24px"
                      viewBox="0 0 24 24"
                      width="24px"
                      fill="rgb(43,166,64)"
                    >
                      <path d="M0 0h24v24H0V0z" fill="none" />
                      <circle cx="15.5" cy="9.5" r="1.5" />
                      <circle cx="8.5" cy="9.5" r="1.5" />
                      <path d="M11.99 2C6.47 2 2 6.48 2 12s4.47 10 9.99 10C17.52 22 22 17.52 22 12S17.52 2 11.99 2zM12 20c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8zm4.41-6.11c-.35-.22-.82-.11-1.03.24-.74 1.17-2 1.87-3.38 1.87s-2.64-.7-3.38-1.88c-.22-.35-.68-.46-1.03-.24-.35.22-.46.68-.24 1.03C8.37 16.54 10.1 17.5 12 17.5s3.63-.97 4.65-2.58c.22-.35.11-.81-.24-1.03z" />
                    </svg>
                  </Avatar>
                }
                label={intl.formatMessage({...messages.selectedFilter})}
                onDelete={() => props.onDelete('participating')}
              />
            </ListItem>
          )}
          {props.filters.notParticipating && (
            <ListItem>
              <Chip
                avatar={
                  <Avatar>
                    <svg
                      xmlns="http://www.w3.org/2000/svg"
                      height="24px"
                      viewBox="0 0 24 24"
                      width="24px"
                      fill="rgb(229,5,69)"
                    >
                      <path d="M0 0h24v24H0V0z" fill="none" />
                      <circle cx="15.5" cy="9.5" r="1.5" />
                      <circle cx="8.5" cy="9.5" r="1.5" />
                      <path d="M11.99 2C6.47 2 2 6.48 2 12s4.47 10 9.99 10C17.52 22 22 17.52 22 12S17.52 2 11.99 2zM12 20c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8zm0-6c-1.9 0-3.63.97-4.65 2.58-.22.35-.11.81.24 1.03.35.22.81.11 1.03-.24.74-1.18 2-1.88 3.38-1.88s2.64.7 3.38 1.88c.14.23.39.35.64.35.14 0 .27-.04.4-.11.35-.22.46-.68.24-1.03C15.63 14.96 13.9 14 12 14z" />
                    </svg>
                  </Avatar>
                }
                label={intl.formatMessage({...messages.notSelectedFilter})}
                onDelete={() => props.onDelete('notParticipating')}
              />
            </ListItem>
          )}
        </div>
      )}
    </div>
  );
};

FilterChips.propTypes = {
  onDelete: PropTypes.func,
  filters: PropTypes.object,
};

export default FilterChips;
