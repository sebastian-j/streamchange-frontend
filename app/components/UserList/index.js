import React from 'react';
import styled from 'styled-components';
import { FormattedMessage } from 'react-intl';
import Button from '@material-ui/core/Button';
import StyledTextField from '../StyledTextField';
import db from '../YoutubeWorker/db';
import UserItem from './userItem';
import PanelTitle from '../Panel/PanelTitle';
import messages from './messages';

const UserListPanel = styled.div`
  background-color: ${props => props.theme.panelBackground};
  display: flex;
  flex-direction: column;
  flex-basis: 0;
  flex-grow: 1;
  margin: 15px;
  padding: 15px;
`;

const Ul = styled.ul`
  overflow-y: scroll;
  list-style: none;
  padding: 0;
`;

const StyledButton = styled(Button)`
  span {
    color: ${props => props.theme.materialButtonColor};
  }
`;

const Counts = styled.span`
  color: ${props => props.theme.secondaryTextColor};
  font-size: 0.9rem;
`;

export default class UserList extends React.Component {
  constructor(props) {
    super(props);
    this.state = { search: '', interval: null, items: [] };
    this.getUsers = this.getUsers.bind(this);
    this.handleInputValueChange = this.handleInputValueChange.bind(this);
    this.toggleEligible = this.toggleEligible.bind(this);
  }

  componentDidMount() {
    this.getUsers();
    this.state.interval = setInterval(this.getUsers.bind(this), 3000);
  }

  getUsers() {
    db.table('users')
      .filter(user =>
        user.title.toLowerCase().includes(this.state.search.toLowerCase()),
      )
      .toArray()
      .then(items => {
        this.setState({ items });
      });
  }

  componentWillUnmount() {
    clearInterval(this.state.interval);
  }

  handleInputValueChange(event) {
    const { target } = event;
    const value = target.type === 'checkbox' ? target.checked : target.value;
    const { name } = target;

    if (value.length < 140) {
      this.setState(
        {
          [name]: value,
        },
        () => this.getUsers(),
      );
    }
  }

  clearList() {
    db.messages.clear();
    db.users.clear().then(() => this.getUsers());
  }

  toggleEligible(id) {
    db.users
      .where('id')
      .equals(id)
      .first()
      .then(user => {
        db.table('users')
          .where('id')
          .equals(id)
          .modify({
            isEligible: !user.isEligible,
          })
          .then(() => {
            this.getUsers();
          });
      });
  }

  render() {
    return (
      <UserListPanel>
        <PanelTitle>
          <FormattedMessage {...messages.panelTitle} />
        </PanelTitle>
        <FormattedMessage {...messages.searchPlaceholder}>
          {placeholder => (
            <StyledTextField
              autoFocus
              margin="dense"
              name="search"
              onChange={this.handleInputValueChange}
              label={placeholder}
              type="text"
              value={this.state.search}
              fullWidth
            />
          )}
        </FormattedMessage>
        <Ul>
          {this.state.items.map(item => (
            <UserItem
              key={item.id}
              channelId={item.id}
              title={item.title}
              isModerator={item.isModerator}
              isSponsor={item.isSponsor}
              isEligible={item.isEligible}
              handleToggleUser={this.toggleEligible}
            />
          ))}
        </Ul>
        <StyledButton onClick={this.clearList}>
          <FormattedMessage {...messages.clearBtn} />
        </StyledButton>
        <Counts>
          <FormattedMessage
            {...messages.counter}
            values={{
              selected: this.state.items.filter(
                item => item.isEligible === true,
              ).length,
              all: this.state.items.length,
            }}
          />
        </Counts>
      </UserListPanel>
    );
  }
}
