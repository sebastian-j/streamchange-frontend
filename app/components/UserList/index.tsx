import React, { useEffect, useState } from 'react';
import { FormattedMessage } from 'react-intl';
import { connect } from 'react-redux';
import { createStructuredSelector } from 'reselect';
import IconButton from '@mui/material/IconButton';
import ListItemIcon from '@mui/material/ListItemIcon';
import MenuItem from '@mui/material/MenuItem';
import MenuList from '@mui/material/Menu';
import Tooltip from '@mui/material/Tooltip';

import db from '../YoutubeWorker/db';
import { useInjectReducer } from '../../utils/injectReducer';
import { Counts } from './components/Counts';
import FilterChips from './FilterChips';
import { Header } from './components/Header';
import { HeaderButtons } from './components/HeaderButtons';
import { StyledButton } from './components/StyledButton';
import StyledTextField from '../StyledTextField';
import { ThemedSvg } from './components/ThemedSvg';
import UserItem from './userItem';
import { UserListPanel } from '../Panel/UserListPanel';
import PanelTitle from '../Panel/PanelTitle';
import { makeSelectUserArray } from './selectors';
import { makeSelectGiveawayRequirement } from '../GiveawayRules/selectors';
import {
  deselectAllUsers,
  getListFromIdb,
  purgeList,
  selectAllUsers,
  toggleEligibility,
} from './actions';
import messages from './messages';
import reducer from './reducer';
import { FilteringOptions, User } from './types';

interface Props {
  deselectAllUsers: () => void;
  getList: (arr: User[]) => void;
  giveawayReq: number;
  purgeList: () => void;
  selectAllUsers: () => void;
  toggleEligibility: (id: string) => void;
  userArray: Array<User>;
}

const UserList = (props: Props) => {
  useInjectReducer({ key: 'userList', reducer });
  const [anchorEl, setAnchorEl] = useState<Element | null>(null);
  const [filtersAnchorEl, setFiltersAnchorEl] = useState<Element | null>(null);
  const [searchQuery, setSearchQuery] = useState<string>('');
  const [items, setItems] = useState<Array<User>>([]);
  const [filters, setFilters] = useState<FilteringOptions>({
    moderators: false,
    sponsors: false,
    verified: false,
    regulars: false,
    participating: false,
    notParticipating: false,
  });
  let selectedCount = 0;
  let allCount = 0;

  const openMenu = (event) => {
    setAnchorEl(event.currentTarget);
  };

  const closeMenu = () => {
    setAnchorEl(null);
  };

  const openFiltersMenu = (event) => {
    setFiltersAnchorEl(event.currentTarget);
  };

  const closeFiltersMenu = () => {
    setFiltersAnchorEl(null);
  };

  const handleInputValueChange = (event) => {
    const { target } = event;
    const value = target.type === 'checkbox' ? target.checked : target.value;
    setSearchQuery(value);
    if (value.length > 0 && value.length < 140) {
      setItems(
        props.userArray.filter((item) =>
          item.title.toLowerCase().includes(value.toLowerCase()),
        ),
      );
    } else if (value.length === 0) {
      setItems([]);
    }
  };

  const isFiltering = () => Object.values(filters).some((x) => x);

  const getUsers = (): Array<User> => {
    let ret: Array<User> = items.length === 0 ? props.userArray : items;
    if (props.giveawayReq === 1)
      ret = ret.filter((user) => user.isSponsor !== false);
    if (isFiltering()) {
      if (filters.participating && !filters.notParticipating) {
        ret = ret.filter((user) => user.isEligible);
      } else if (!filters.participating && filters.notParticipating) {
        ret = ret.filter((user) => !user.isEligible);
      }
      if (
        filters.moderators ||
        filters.regulars ||
        filters.sponsors ||
        filters.verified
      ) {
        ret = ret.filter(
          (user) =>
            (filters.moderators && user.isModerator) ||
            (filters.sponsors && user.isSponsor) ||
            (filters.verified && user.isVerified) ||
            (filters.regulars && !user.isModerator && !user.isSponsor),
        );
      }
    }
    selectedCount = ret.filter((item) => item.isEligible).length;
    allCount = ret.length;
    return ret;
  };

  useEffect(() => {
    if (props.userArray.length === 0) {
      db.table('users')
        .toArray()
        .then((arr) => {
          props.getList(arr);
        });
    }
  }, []);

  return (
    <UserListPanel>
      <Header>
        <PanelTitle>
          <FormattedMessage {...messages.panelTitle} />
        </PanelTitle>
        <HeaderButtons>
          <FormattedMessage {...messages.filtersTooltip}>
            {(title) => (
              <Tooltip title={title} aria-label="add">
                <IconButton
                  edge="end"
                  aria-label="Options"
                  onClick={openFiltersMenu}
                >
                  <ThemedSvg
                    xmlns="http://www.w3.org/2000/svg"
                    viewBox="0 0 24 24"
                    width="18px"
                    height="18px"
                  >
                    <g>
                      <path d="M0,0h24 M24,24H0" fill="none" />
                      <path d="M4.25,5.61C6.27,8.2,10,13,10,13v6c0,0.55,0.45,1,1,1h2c0.55,0,1-0.45,1-1v-6c0,0,3.72-4.8,5.74-7.39 C20.25,4.95,19.78,4,18.95,4H5.04C4.21,4,3.74,4.95,4.25,5.61z" />
                      <path d="M0,0h24v24H0V0z" fill="none" />
                    </g>
                  </ThemedSvg>
                </IconButton>
              </Tooltip>
            )}
          </FormattedMessage>
          <IconButton edge="end" aria-label="Options" onClick={openMenu}>
            <ThemedSvg
              xmlns="http://www.w3.org/2000/svg"
              viewBox="0 0 24 24"
              width="18px"
              height="18px"
            >
              <path d="M0 0h24v24H0z" fill="none" />
              <path d="M12 8c1.1 0 2-.9 2-2s-.9-2-2-2-2 .9-2 2 .9 2 2 2zm0 2c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2zm0 6c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2z" />
            </ThemedSvg>
          </IconButton>
        </HeaderButtons>
        <MenuList
          id="options-menu"
          anchorEl={anchorEl}
          open={!!anchorEl}
          onClose={closeMenu}
        >
          <MenuItem
            onClick={() => {
              props.selectAllUsers();
              closeMenu();
            }}
          >
            <FormattedMessage {...messages.selectUsers} />
          </MenuItem>
          <MenuItem
            onClick={() => {
              props.deselectAllUsers();
              closeMenu();
            }}
          >
            <FormattedMessage {...messages.deselectUsers} />
          </MenuItem>
        </MenuList>
        <MenuList
          id="filters-menu"
          anchorEl={filtersAnchorEl}
          open={!!filtersAnchorEl}
          onClose={closeFiltersMenu}
        >
          <MenuItem
            onClick={() =>
              setFilters((prevState) => ({
                ...prevState,
                regulars: !prevState.regulars,
              }))
            }
          >
            <ListItemIcon>
              {filters.regulars && (
                <svg
                  focusable="false"
                  viewBox="0 0 24 24"
                  aria-hidden="true"
                  height="25px"
                >
                  <path d="M9 16.17L4.83 12l-1.42 1.41L9 19 21 7l-1.41-1.41z" />
                </svg>
              )}
            </ListItemIcon>
            <FormattedMessage {...messages.regularsFilter} />
          </MenuItem>
          <MenuItem
            onClick={() =>
              setFilters((prevState) => ({
                ...prevState,
                moderators: !prevState.moderators,
              }))
            }
          >
            <ListItemIcon>
              {filters.moderators && (
                <svg
                  focusable="false"
                  viewBox="0 0 24 24"
                  aria-hidden="true"
                  height="25px"
                >
                  <path d="M9 16.17L4.83 12l-1.42 1.41L9 19 21 7l-1.41-1.41z" />
                </svg>
              )}
            </ListItemIcon>
            <FormattedMessage {...messages.moderatorsFilter} />
          </MenuItem>
          <MenuItem
            onClick={() =>
              setFilters((prevState) => ({
                ...prevState,
                sponsors: !prevState.sponsors,
              }))
            }
          >
            <ListItemIcon>
              {filters.sponsors && (
                <svg
                  focusable="false"
                  viewBox="0 0 24 24"
                  aria-hidden="true"
                  height="25px"
                >
                  <path d="M9 16.17L4.83 12l-1.42 1.41L9 19 21 7l-1.41-1.41z" />
                </svg>
              )}
            </ListItemIcon>
            <FormattedMessage {...messages.sponsorsFilter} />
          </MenuItem>
          <MenuItem
            onClick={() =>
              setFilters((prevState) => ({
                ...prevState,
                verified: !prevState.verified,
              }))
            }
          >
            <ListItemIcon>
              {filters.verified && (
                <svg
                  focusable="false"
                  viewBox="0 0 24 24"
                  aria-hidden="true"
                  height="25px"
                >
                  <path d="M9 16.17L4.83 12l-1.42 1.41L9 19 21 7l-1.41-1.41z" />
                </svg>
              )}
            </ListItemIcon>
            <FormattedMessage {...messages.verifiedFilter} />
          </MenuItem>
          <MenuItem
            onClick={() =>
              setFilters((prevState) => ({
                ...prevState,
                participating: !prevState.participating,
              }))
            }
          >
            <ListItemIcon>
              {filters.participating && (
                <svg
                  focusable="false"
                  viewBox="0 0 24 24"
                  aria-hidden="true"
                  height="25px"
                >
                  <path d="M9 16.17L4.83 12l-1.42 1.41L9 19 21 7l-1.41-1.41z" />
                </svg>
              )}
            </ListItemIcon>
            <FormattedMessage {...messages.selectedFilter} />
          </MenuItem>
          <MenuItem
            onClick={() =>
              setFilters((prevState) => ({
                ...prevState,
                notParticipating: !prevState.notParticipating,
              }))
            }
          >
            <ListItemIcon>
              {filters.notParticipating && (
                <svg
                  focusable="false"
                  viewBox="0 0 24 24"
                  aria-hidden="true"
                  height="25px"
                >
                  <path d="M9 16.17L4.83 12l-1.42 1.41L9 19 21 7l-1.41-1.41z" />
                </svg>
              )}
            </ListItemIcon>
            <FormattedMessage {...messages.notSelectedFilter} />
          </MenuItem>
        </MenuList>
      </Header>
      <FilterChips
        filters={filters}
        onDelete={(key) =>
          setFilters((prevState) => ({
            ...prevState,
            [key]: !prevState[key],
          }))
        }
      />
      <FormattedMessage {...messages.searchPlaceholder}>
        {(placeholder) => (
          <StyledTextField
            autoFocus
            margin="dense"
            name="search"
            onChange={handleInputValueChange}
            label={placeholder}
            type="text"
            value={searchQuery}
            variant="standard"
            fullWidth
          />
        )}
      </FormattedMessage>
      <ul>
        {getUsers().map((item) => (
          <UserItem
            key={item.id}
            channelId={item.id}
            title={item.title}
            isModerator={item.isModerator}
            isSponsor={item.isSponsor}
            isVerified={item.isVerified}
            isEligible={item.isEligible}
            handleToggleUser={props.toggleEligibility}
          />
        ))}
      </ul>
      <StyledButton onClick={props.purgeList} color="inherit">
        <span>
          <FormattedMessage {...messages.clearBtn} />
        </span>
      </StyledButton>
      <Counts>
        <FormattedMessage
          {...messages.counter}
          values={{
            selected: selectedCount,
            all: allCount,
          }}
        />
      </Counts>
    </UserListPanel>
  );
};

export function mapDispatchToProps(dispatch) {
  return {
    deselectAllUsers: () => dispatch(deselectAllUsers()),
    getList: (arr) => dispatch(getListFromIdb(arr)),
    purgeList: () => dispatch(purgeList()),
    selectAllUsers: () => dispatch(selectAllUsers()),
    toggleEligibility: (id) => dispatch(toggleEligibility(id)),
    dispatch,
  };
}

const mapStateToProps = createStructuredSelector({
  giveawayReq: makeSelectGiveawayRequirement(),
  userArray: makeSelectUserArray(),
});

export default connect(mapStateToProps, mapDispatchToProps)(UserList);
