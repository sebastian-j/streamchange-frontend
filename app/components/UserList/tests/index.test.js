import React from 'react';
import { Provider } from 'react-redux';
import ShallowRenderer from 'react-test-renderer/shallow';
import configureStore from 'redux-mock-store';

import UserList from '../index';

const renderer = new ShallowRenderer();
const mockStore = configureStore([]);

describe('<UserList />', () => {
  let store;
  beforeEach(() => {
    store = mockStore({
      userArray: [
        { id: 'id1', title: 'user1' },
        { id: 'id2', title: 'user2' },
      ],
    });
  });
  it('should render and match the snapshot', () => {
    renderer.render(
      <Provider store={store}>
        <UserList />
      </Provider>,
    );
    const renderedOutput = renderer.getRenderOutput();
    expect(renderedOutput).toMatchSnapshot();
  });
});
