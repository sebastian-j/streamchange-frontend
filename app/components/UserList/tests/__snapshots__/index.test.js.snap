// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<UserList /> should render and match the snapshot 1`] = `
<UserList__UserListPanel>
  <PanelTitle>
    <FormattedMessage
      defaultMessage="Users"
      id="streamchange.components.UserList.title"
      values={Object {}}
    />
  </PanelTitle>
  <FormattedMessage
    defaultMessage="Search"
    id="streamchange.components.UserList.search.placeholder"
    values={Object {}}
  >
    [Function]
  </FormattedMessage>
  <UserList__Ul />
  <UserList__StyledButton
    onClick={[Function]}
  >
    <FormattedMessage
      defaultMessage="Clear all"
      id="streamchange.components.UserList.clearBtn"
      values={Object {}}
    />
  </UserList__StyledButton>
  <UserList__Counts>
    <FormattedMessage
      defaultMessage="{selected} out of {all} are participating"
      id="streamchange.components.UserList.counter"
      values={
        Object {
          "all": 0,
          "selected": 0,
        }
      }
    />
  </UserList__Counts>
</UserList__UserListPanel>
`;
